### ISSUE TITLE
apt update error for goaccess Ubuntu arm64

### ISSUE THREAD

**Issue by SyntrophicConsortium** [allinurl/goaccess#1894]
I recently installed goaccess arm64 on Ubuntu 20.04.1 using the distro specific instructions here (so it installs a newer version than what is in the official repo). This worked fine, and subsequent `apt update` commands completed execution without errors, but today I noticed this when running `apt update`:

`Skipping acquire of configured file 'main/binary-arm64/Packages' as repository 'https://deb.goaccess.io focal InRelease' doesn't support architecture 'arm64'`
 
/etc/apt/sources.list.d/goaccess.list: 
`deb https://deb.goaccess.io/ focal main`

For the time being, I've just commented out that line, but I thought I would mention it as I didn't see an issue here for it.

**Comment by allinurl** [comment]
Good point. I recently deprovisioned an arm server that was building these releases as I moved to automated vagrant builds. I don't recall finding an arm based image, I may need to look again or fire up again the arm server. I'll keep it open for the time being so I can look further into it. Thanks.

**Comment by 0bi-w6n-K3nobi** [comment]
Can be use Docker or LXC container emulate ARM64 for that.
Or full virtualization with QEmu.
It is similar a vagrand. This is no so hard setup.

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi I agree, I'll be more incline to go the Docker or LXC container. Do you have a reference where I can look for such setup? Thanks!

**Comment by 0bi-w6n-K3nobi** [comment]
@allinurl What is your daily use distro?

**Comment by 0bi-w6n-K3nobi** [comment]
It's great!
I'm use Fedora/Centos a long long time... 
Unfortunately there isn't version for AARCH64 [ARM64]...

**Comment by 0bi-w6n-K3nobi** [comment]
Ok, I got it. You want an Unbuntu Docker/LXC that runs on Arch Linux ...

**Comment by allinurl** [comment]
Wondering if it can be done with github actions... 

> GitHub Actions now supports self-hosted runners on the ARM and ARM64 architectures.
https://github.blog/changelog/2019-12-03-github-actions-self-hosted-runners-on-arm-architectures/

**Comment by 0bi-w6n-K3nobi** [comment]
**Quick Guide: Running non-native arch LXC container under Arch Linux**

All rigth. Let's start from beginning.

As I said earlier, I prefer LXC over Docker.
It is more like a VM with the exception that there is no kernel running.

_Documentation:_
Arch Linux as always has impeccable [documentation](https://wiki.archlinux.org/index.php/Linux_Containers). But it is good to take a look one at [Debian](https://wiki.debian.org/LXC) too.

_Network:_
I always use the network on the guest via bridge with the host.
In Docker it is like a bridge as everything happens through NAT.
In this case, you have two options: `lxc-net` or `libvirtd`.

`lxc-net`
It's simple. It works well, however it conflicts with `dnsmasq` service.
Then it will be necessary to disable `dnsmaq`. [ `libvird` running? ]
[ This is the only solution proposed by the LXC maintainers and various forums ]
Its default netmask is `10.0.3.0/24`. The bridge interface is `lxcbr0`.
It's so easy for lxc container nested. [ Yes, nested ]

`libvirtd`
Works like `lxc-net`. However, your `DNS/DHCP` coexists peacefully with `dnsmasq`.
It is the standard virtualization library that works with `qemu/kvm` and `virt-manager`.
Its netmask is `192.168.222.0/24` and interface `virbr0`. I prefer this. And I use `kvm` too.

_Installation:_
There are three packages: `lxc`, `lxc-templates` and `lxcfs`. The latter is optional.
`lxc-templates` is unfortunately not found in the standard Arch Linux repository.
There is only the [git version](https://aur.archlinux.org/lxc-templates.git). But I managed to compile it and is available [here](https://github.com/allinurl/goaccess/files/5252679/lxc-templates-3-0-4-1.zip).
Also install the packages: `debootstrap`, `dnsmasq` and `which`.
Sorry about that, but the installation dependencies are defective.

_Setup:_
After installation if you are going to use `lxc-net` there is a small touch.
It is necessary to create the file `/etc/default/lxc-net`
[ `/etc/sysconfig/lxc-net` in some distro like Fedora ]

with the following line:

`USE LXC_BRIDGE = "true"`

before starting the service.

_Create the container:_
**The hard way**, or standard form would be:

`lxc-create -n name-of-container -t template-definition -- -option1 -option2 ...`

and for Ubuntu in native architecture it would be:

`lxc-create -n ubuntu-focal -t ubuntu -- -r focal --packages net-tools`

[ and long long long time after ... ] and work like a charm!

However for non-native architecture such as arm64 it would be enough to add some like `-a arm64`.
And then you can see the error message: `can't create arm64 container on amd64` .
But not everything is lost. 
Simply edit the template `/usr/share/lxc/templates/lxc-ubuntu` file and comment out lines `808-811`.
Again so more long time because arm64 is emulated...

**The soft way**, or there is an image repository for LXC. You can see the list of images [here](https://us.images.linuxcontainers.org/).
Then use:

 `lxc-create -t download -n ubuntu-focal-arm64 -- -d ubuntu -r focal -a arm64`
 
This will be quick and clean. Now, we're almost there.
It doesn't matter which of the methods used to create the container. The result is very similar.
Don't forget to change the bridge interface if you use `libvirtd` in `/var/lib/lxc/name-of-container/config`.

_Problem:_
After installation, and this happens for the native version of architecture, just start.
But for non-native archs the `systemd-resolved`, `systemd-networkd` or `NetworkManager` services just don't work !!!
My bet is that the network interface is not exactly the same on all archs.
Remembering here the implementation of the signal. [ see old `man 7 signal`  [here](https://manpages.debian.org/stretch/manpages/signal.7.en.html) ]

The workaround is to use `ifconfig` and `route` to manually adjust the network.
However the installation of the repository image does not come with the `net-tools` package.

_The Solutions:_
**first** - Download the arm64 package from `net-tools` [here](http://turul.canonical.com/pool/main/n/net-tools/net-tools_1.60+git20180626.aebd88e-1ubuntu1_arm64.deb) and install manually.

**second** - the biggest.
If we look here we already have a complete system tree ready for chroot jail. Only the network is missing.
The trick is to use the host's network:

`mount --bind /proc /var/lib/lxc/name-of-container/rootfs/proc`
remove or move `/var/lib/lxc/name-of-container/rootfs/etc/resolv.conf` [ is a link to `systemd-resolved`'s file ]
`cp /etc/resolv.conf /var/lib/lxc/name-of-container/rootfs/etc/resolv.conf`

and finally:
`chroot  /var/lib/lxc/name-of-container/rootfs`
`passwd root`  [ don't forget change password ]

It is even possible to add a user and use sudo for a new one.
If enough, you can use the compiler and generate packages here.
and apt-get will work perfectly. So:

`apt-get install net-tools`

for finish:

`exit`

_Start the Container:_
But if we want an isolated sandbox system:  [ We finally arrived here ... ]

`lxc-start -n name-of-container -F` [ foreground like a vm ]

after login, for `libvirtd`:

`ifconfig eth0 192.168.122.5`  [ of your choice ]
`route add default gw 192.168.122.1`
`echo nameserver 192.168.122.1 >/etc/resolv.conf`

or `10.0.3.X` for `lxc-net`.

Restart `sshd` if you use that.
And everything should work as expected.

_Aditionals and tips:_
As already seen above, 
lxc will create the container under `/var/lib/lxc` and a cache for each `distro/version/arch` under `/var/cache/lxc`.
If you have a large separate `/home` partition, I strong suggest use soft-link for the above mentioned directories for this partition.
i.e. use `/home` for large, volatile and non-system data.

For Fedora/CentOS/RH packages I use `mock`.
It is very functional, allowing 32 and 64 bits, for OS versions above and below the host.
Build a minimal directory tree for compilation, automatically installing additional packages needed.
Always runs on ordinary user.

It is possible to run the LXC container as an ordinary user without privileges ... but that is another story ...
It is also possible to limit the use of CPU, memory and SWAP.

And we come to the end of our journey.
I hope this guide is satisfactory for you.
Good job!

**Comment by 0bi-w6n-K3nobi** [comment]
_PS:_ I forgot this -- `CGroup`.

LXC between version 3 and 4 changed from CGroup V1 to V2.
I believe that Docker still requires the V1 version.
In this case, if you have problems use at boot line:

`systemd.unified_cgroup_hierarchy = 0`

Some documentation on this is found [here](https://wiki.debian.org/LXC/CGroupV2).

**Comment by alqu** [comment]
Hmm, this does not seem to work.
I followed the installation instructions from https://goaccess.io/download#official-repo
At least not on an Apple M1-Machine while trying to install on Docker/Debian 11 I get this result:
```shell
sudo apt-get update
Get:1 http://security.debian.org/debian-security bullseye-security InRelease [44.1 kB]
Hit:2 http://deb.debian.org/debian bullseye InRelease                               
Get:3 http://deb.debian.org/debian bullseye-updates InRelease [39.4 kB]
Get:4 https://deb.goaccess.io bullseye InRelease [1631 B]  
Fetched 85.1 kB in 5s (18.7 kB/s)  
Reading package lists... Done
N: Skipping acquire of configured file 'main/binary-arm64/Packages' as repository 'https://deb.goaccess.io bullseye InRelease' doesn't support architecture 'arm64'
```

Some info from that machine:
```
uname -a
Linux goaccesstesthost 5.10.104-linuxkit #1 SMP PREEMPT Thu Mar 17 17:05:54 UTC 2022 aarch64 GNU/Linux
```

I do not know, what causes this error since your list above clearly lists this architecture.

**Comment by alqu** [comment]
Oops, answering my own question: obviously there is the combination bullseye + arm64 missing. Doh.

**Comment by allinurl** [comment]
@alqu I'll need to revisit this as bullseye arm64 was segfaulting under qemu when running apt-get update. Let me check again and will post back.

**Comment by alqu** [comment]
@allinurl any news on the image? :-)

**Comment by allinurl** [comment]
@alqu Please give it a shot and let me know how it goes - just added arm64 bullseye. Here's the latest list:
```

goaccess |      1.6.1-jammy |    jammy | arm64
goaccess |    2:1.6.1-jammy |    jammy | amd64
goaccess |   1:1.6.1-impish |   impish | amd64
goaccess |   1:1.6.1-bionic |   bionic | arm64, amd64
goaccess |     1.6.1-bionic |   bionic | i386
goaccess |    1:1.6.1-focal |    focal | arm64, amd64
goaccess |     1.6.1-xenial |   xenial | arm64, amd64, i386
goaccess |    1.6.1-stretch |  stretch | arm64
goaccess |  1:1.6.1-stretch |  stretch | amd64, i386
goaccess |   2:1.6.1-buster |   buster | arm64, amd64
goaccess |   1.6.1-bullseye | bullseye | arm64, i386
goaccess | 1:1.6.1-bullseye | bullseye | amd64

```

**Comment by alqu** [comment]
@allinurl Awesome. It works perfectly. Thanks a lot!

**Comment by allinurl** [comment]
armhf is now supported as well.
```
goaccess |    2:1.6.2-jammy |    jammy | armhf, arm64, amd64
goaccess |   1:1.6.2-impish |   impish | amd64
goaccess |   1:1.6.2-bionic |   bionic | armhf, arm64, amd64, i386
goaccess |    1:1.6.2-focal |    focal | arm64, amd64
goaccess |   1:1.6.2-xenial |   xenial | armhf, arm64, amd64, i386
goaccess |  1:1.6.2-stretch |  stretch | armhf, arm64, amd64, i386
goaccess |   2:1.6.2-buster |   buster | arm64, amd64
goaccess | 1:1.6.2-bullseye | bullseye | armhf, arm64, amd64, i386
```
